reactioncommerce/reaction
# Feat 818 prinzdezibel logger fatal send email

Resolves #818   
Type: **feature**

This PR closes #818

## Issue
We want to notify sysadmin if we hit a fatal error so that they can get their site back up.

## Note
This PR has a companion PR in repository @reactioncommerce/logger

## Solution / Changes

This PR implements the minimal requirements listed in #818: Sending email alerts to a admin whenever a Logger.fatal() is executed. There are many possible ways to enhance this through more sophisticated handling (e.g. persistency, ensure that no duplicates are send, etc, etc). This is just intended as starting point.

- Use a mutuable export to allow a email stream to be attached to the
 logger instance after the application has been intialized (afterCoreInit). This is necessary to have the environment in place to specify nodemailer's SMTP options

- Create a new email bunyan stream for sending fatal errors as email

- Introduce a new environment variable REACTION_ENABLE_ALERT_NOTFICATIONS
  to enable email notifications. In the default setup, the emails get sent to the shop owner's email address

- Introduce a new environment variable REACTION_ALERT_NOTFICATIONS_EMAIL to override the recipient email address.

## Breaking changes
None expected

## Testing
1. Make sure you've the code from the companion PR in @reactioncommerce/logger in place.
This can be tricky if you are intend to incorporate the logger NPM package's code as linked NPM package (as opposed to npm install a published version of it). This is because @reactioncommerce/logger now contains a **peer dependency** (@reactioncommerce/hooks), which is known to be a bit difficult to deal with. See also: http://codetunnel.io/you-can-finally-npm-link-packages-that-contain-peer-dependencies/

TLDR: To test it, you may just copy over the source code from @reactioncommerce/logger to the main project's node_module folder, e.g.
```
cd ~/github/reactioncommerce/reaction/node_modules
cp -Rav ~/github/reactioncommerce/logger .
```

2. Add somewhere in server side code, e.g. in  "cart/addToCart" meteor method
`Logger.fatal("email notification test");`

3. Run reaction with proper SMTP settings, e.g. reaction rc-config
```
export REACTION_ENABLE_ALERT_NOTFICATIONS=1 && export REACTION_ALERT_NOTFICATIONS_EMAIL="<youremail>@reactioncommerce.com" && reaction --settings settings/settings.json
```
4. Navigate to the example product and put it to cart
5. See the fatal error message in the server console
6. Go to your email inbox and see an incoming email containing that error
